dbms/src/DataStreams/RuntimeFilter.cpp

```diff
@@ -0,0 +1,178 @@
+{
+    source_column_name = source_column_name_;
+}
+
+void RuntimeFilter::setINValuesSet(const std::shared_ptr<Set> & in_values_set_)
+{
+    in_values_set = in_values_set_;
+}
+
+// There is a possibility that the same rf is canceled multiple times,
+// but rf does not have thread safety issues.
+void RuntimeFilter::updateValues(const ColumnWithTypeAndName & values, const LoggerPtr & log)
+{
+    if (isFailed())
+    {
+        return;
+    }
+    Block data(std::initializer_list<ColumnWithTypeAndName>({values}));
+    switch (rf_type)
+    {
+    case tipb::IN:
+        try
+        {
+            LOG_DEBUG(log, "update values in rf, data size:{}", data.rows());
+            in_values_set->insertFromBlock(data, true);
+        }
+        catch (Exception & e)
+        {
+            if (e.code() == ErrorCodes::SET_SIZE_LIMIT_EXCEEDED)
+            {
+                updateStatus(RuntimeFilterStatus::FAILED, "The rf in values exceed the limit. " + e.message());
+                LOG_WARNING(log, "cancel runtime filter id:{}, reason: {} ", id, failed_reason);
+            }
+        }
+        break;
+    case tipb::MIN_MAX:
+        // todo
+        break;
+    }
+}
+
+void RuntimeFilter::finalize(const LoggerPtr & log)
+{
+    if (isFailed() || isReady())
+    {
+        return;
+    }
+    updateStatus(RuntimeFilterStatus::READY);
+    std::string rf_values_info;
+    switch (rf_type)
+    {
+    case tipb::IN:
+        rf_values_info = fmt::format("number of IN values:{}", in_values_set->getTotalRowCount());
+        break;
+    case tipb::MIN_MAX:
+        // TODO
+        break;
+    }
+    LOG_INFO(log, "finalize runtime filter id:{}, rf values info:{}", id, rf_values_info);
+}
+
+void RuntimeFilter::cancel(const LoggerPtr & log, const std::string & reason)
+{
+    if (isFailed() || isReady())
+    {
+        return;
+    }
+    updateStatus(RuntimeFilterStatus::FAILED, reason);
+    LOG_INFO(log, "cancel runtime filter id:{}, reason:{}", id, reason);
+}
+
+bool RuntimeFilter::isReady()
+{
+    return status == RuntimeFilterStatus::READY;
+}
+
+bool RuntimeFilter::isFailed()
+{
+    return status == RuntimeFilterStatus::FAILED;
+}
+
+bool RuntimeFilter::await(int64_t ms_remaining)
+{
+    if (isFailed())
+    {
+        return false;
+    }
+    if (!isReady())
+    {
+        if (ms_remaining <= 0)
+        {
+            return isReady();
+        }
+        std::unique_lock<std::mutex> lock(inner_mutex);
+        return inner_cv.wait_for(lock, std::chrono::milliseconds(ms_remaining), [this] { return isReady(); });
+    }
+    return true;
+}
+
+void RuntimeFilter::updateStatus(RuntimeFilterStatus status_, const std::string & reason)
+{
+    switch (status_)
+    {
+    case RuntimeFilterStatus::NOT_READY:
+        // The initial state of the runtime filter is not ready. It can only be updated as FAILED or READY.
+        return;
+    case RuntimeFilterStatus::FAILED:
+        failed_reason = reason;
```

Comment:
Still no need, if you want to ensure this, you need to add a lock to control the logic of ```checkAndUpdate```. At present, only ```failed_reasion``` is non-thread-safe and does not affect the overall code logic, and there is no need to lock it. This achieves better efficiency.

Reply:
But concurrently write to a string may cause some other unknown problems. And a side effect is the first failed reason will be overwritten by the later one, which is also not expected.

Reply:
You can refer to this pr(https://github.com/pingcap/tiflash/pull/2421), the original problem is TiFlash crash randomly when meet error, but the crash is weird because crash stack is different and looks like randomly for each crash. It takes us a long time to figure out the root cause is actually a concurrent write/read to `ExchangeReceiver::err`

Reply:
Looks in this PR(#2421). ```err_msg``` still has the problem of reading and writing at the same time. 
Read thread: 
![image](https://user-images.githubusercontent.com/111482174/236432165-24a3bb8d-a6ec-4e09-a5b5-de180addc1c1.png)
Write thread:
![image](https://user-images.githubusercontent.com/111482174/236432290-c9716a3c-fe23-42be-b66d-26b3673339b3.png)

Just changed the exception object to a string object.
Then I also use the string object here will not be a problem?

Reply:
Sorry, it is my fault. I saw ```std::lock_guard<std::mutex> lock(mu);``` logic before you update err_msg.
I will change the string object to thread safe.